ethcore/src/header.rs

```diff
@@ -75,14 +75,18 @@ pub struct Header {
 	/// Vector of post-RLP-encoded fields.
 	seal: Vec<Bytes>,
 
-	/// The memoized hash of the RLP representation *including* the seal fields.
-	hash: RefCell<Option<H256>>,
-	/// The memoized hash of the RLP representation *without* the seal fields.
-	bare_hash: RefCell<Option<H256>>,
+	/// Memoized hash of that header and the seal.
+	hash: Option<H256>,
 }
 
 impl PartialEq for Header {
 	fn eq(&self, c: &Header) -> bool {
+		if let (&Some(ref h1), &Some(ref h2)) = (&self.hash, &c.hash) {
+			if h1 == h2 {
```

Comment:
Shouldn't this be `return h1 == h2` instead? I.e. if the hash is defined for both we do the comparison strictly on the hash.